jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/AuthenticatePushStep.java

```diff
@@ -21,58 +21,53 @@
 import com.google.cloud.tools.jib.builder.ProgressEventDispatcher;
 import com.google.cloud.tools.jib.builder.TimerEventDispatcher;
 import com.google.cloud.tools.jib.configuration.BuildContext;
-import com.google.cloud.tools.jib.http.Authorization;
-import com.google.cloud.tools.jib.registry.RegistryAuthenticator;
+import com.google.cloud.tools.jib.registry.RegistryClient;
+import com.google.cloud.tools.jib.registry.credentials.CredentialRetrievalException;
 import java.io.IOException;
-import java.util.Optional;
 import java.util.concurrent.Callable;
-import javax.annotation.Nullable;
 
 /**
  * Authenticates push to a target registry using Docker Token Authentication.
  *
  * @see <a
  *     href="https://docs.docker.com/registry/spec/auth/token/">https://docs.docker.com/registry/spec/auth/token/</a>
  */
-class AuthenticatePushStep implements Callable<Optional<Authorization>> {
+class AuthenticatePushStep implements Callable<RegistryClient> {
 
   private static final String DESCRIPTION = "Authenticating push to %s";
 
   private final BuildContext buildContext;
   private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
-  @Nullable private final Credential registryCredential;
 
   AuthenticatePushStep(
-      BuildContext buildContext,
-      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
-      @Nullable Credential registryCredential) {
+      BuildContext buildContext, ProgressEventDispatcher.Factory progressEventDispatcherFactory) {
     this.buildContext = buildContext;
     this.progressEventDispatcherFactory = progressEventDispatcherFactory;
-    this.registryCredential = registryCredential;
   }
 
   @Override
-  public Optional<Authorization> call() throws IOException, RegistryException {
+  public RegistryClient call() throws CredentialRetrievalException, IOException, RegistryException {
     String registry = buildContext.getTargetImageConfiguration().getImageRegistry();
     try (ProgressEventDispatcher ignored =
             progressEventDispatcherFactory.create("authenticating push to " + registry, 1);
         TimerEventDispatcher ignored2 =
             new TimerEventDispatcher(
                 buildContext.getEventHandlers(), String.format(DESCRIPTION, registry))) {
-      Optional<RegistryAuthenticator> registryAuthenticator =
+      Credential credential =
+          RegistryCredentialRetriever.getTargetImageCredential(buildContext).orElse(null);
+
+      RegistryClient registryClient =
           buildContext
               .newTargetImageRegistryClientFactory()
-              .newRegistryClient()
-              .getRegistryAuthenticator();
-      if (registryAuthenticator.isPresent()) {
-        return Optional.of(registryAuthenticator.get().authenticatePush(registryCredential));
+              .setCredential(credential)
```

Comment:
I find this flow to be a little strange. But maybe that's just a consequence of our code, should the registry client just handle this basic auth stuff? 

Is there a reason we want the steps of authorizing the client to exist separate and outside the client?

Reply:
Actually looking at stuff further down the code, the flow is ever so slightly different everywhere?

Reply:
Yeah. I understand the flow may seem strange and make you think perhaps logic could just be put inside the registry client. I thought the same.

I don't have any history, but our way of doing auth has been like this:

- For pull (base image): first try without any credentials. If it fails, configure basic auth no matter what and try again. If it fails, initiate pull bearer auth.
- For push (target image): always initiate push bearer auth first. If the server returns asking for basic auth, configure basic auth.

So, for now, basically the overall auth logic is different per use-case and needs to be defined at a higher-level per. That's why I made the registry client provides the necessary knobs (`configureBasicAuth` and `doPull/PushBearerAuth`) and let the caller decided how to drive it.